### PR TITLE
HTBHF-423 Added address entity to the claimant.

### PR DESCRIPTION
**Pull request by dwybourn**

### CODE REVIEW

File: db/src/main/resources/db/migration/V1_007__add_address.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+create table address (
+    id uuid not null primary key,
+    address_line_1 varchar(500) not null,
+    address_line_2 varchar(500),
+    town_or_city varchar(500) not null,
+    postcode varchar(7) not null
+);
+
+alter table claimant add column address_id uuid;
```
Comment (YetAnotherMatt) [inline db/src/main/resources/db/migration/V1_007__add_address.sql]: might want to be specific about what address this is = e.g. card_delivery_address_id

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    @Column(name = "address_line_1")
+    private String addressLine1;
+
+    @Size(max = 500)
+    @Column(name = "address_line_2")
+    private String addressLine2;
+
+    @NotNull
+    @Size(min = 1, max = 500)
+    @Column(name = "townOrCity")
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java]: Column name doesn't match the format of other column names (underscores)

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+public class AddressTest {
+
+    // Create a string 501 characters long
+    private static final String LONG_STRING = CharBuffer.allocate(501).toString().replace('\0', 'A');
+
+    private Validator validator;
+
+    @BeforeEach
+    void setup() {
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java]: Might it be worthwhile having a `AbstractValidationTest` and move this into there?

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        //Given
+        var address = anAddressWithAddressLine1(LONG_STRING);
+        //When
+        Set<ConstraintViolation<Address>> violations = validator.validate(address);
+        //Then
+        assertThat(violations).hasSingleConstraintViolation("size must be between 1 and 500", "addressLine1");
+    }
+
+    @Test
+    void shouldFailToValidateAddressWhenAddressLine2IsTooLong() {
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java]: I'd also add a test to make sure that a null addressLine2 is valid.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        //Given
+        var address = anAddressWithPostcode(null);
+        //When
+        Set<ConstraintViolation<Address>> violations = validator.validate(address);
+        //Then
+        assertThat(violations).hasSingleConstraintViolation("must not be null", "postcode");
+    }
+
+    @Test
+    void shouldFailToValidateAddressWithInvalidPostcode() {
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java]: Can we have a parameterised test with lots of values here please rather than just a single one? Blank String, too long, too short, invalid characters etc etc.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/AddressTestDataFactory.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
+package uk.gov.dhsc.htbhf.claimant.testsupport;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import uk.gov.dhsc.htbhf.claimant.entity.Address;
+
+/**
+ * Test data factory for {@link uk.gov.dhsc.htbhf.claimant.entity.Address} objects.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/AddressTestDataFactory.java]: Pointless??

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java (hunk 2 above)
Comment (mikeSmith08) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java]: In fact the column name doesn't match, we should have a failing unit test for this...

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Claimant.java [hunk 7]
```diff
@@ -52,6 +55,9 @@
     @Column(name = "date_of_birth")
     private LocalDate dateOfBirth;
 
+    @OneToOne(mappedBy = "claimant", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Claimant.java]: Do we want this to be lazy loaded, wouldn't we normally want the address available straight away too? I'm not sure, its just a question on whether we should load it at the same time as the Claimant we'd normally need it.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java (hunk 2 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java]: I've updated the name but uncertain as to why the repository test passed. Potentially some hibernate magic or h2 mismatch? Once I've added the api layer in I'll see if this breaks the application.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        //Given
+        var address = anAddressWithPostcode(null);
+        //When
+        Set<ConstraintViolation<Address>> violations = validator.validate(address);
+        //Then
+        assertThat(violations).hasSingleConstraintViolation("must not be null", "postcode");
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"AA1122BB", "A", "11AA21", "", "E!", "EA123"})
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java]: Can we have a blank String in here too please?

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/AddressTestDataFactory.java (hunk 6 above)
Comment (mikeSmith08) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/AddressTestDataFactory.java]: I really do think this is pointless by the way...

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java (hunk 8 above)
Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/entity/AddressTest.java]: There is, fourth argument.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/AddressTestDataFactory.java (hunk 6 above)
Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/AddressTestDataFactory.java]: We can remove it as a pmd rule. @YetAnotherMatt are you ok with that?

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+@Data
+@Builder
+@Table(name = "address")
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+@AllArgsConstructor
+public class Address {
+
+    /**
+     * Regex for matching UK postcodes matching BS7666 format.
+     * { @see https://www.gov.uk/government/publications/bulk-data-transfer-for-sponsors-xml-schema }
```
Comment (YetAnotherMatt) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java]: Is it worth mentioning that the file with the regex definition is BulkDataCommon-v2.1.xsd?

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+     * { @see https://stackoverflow.com/questions/164979/uk-postcode-regex-comprehensive }
+     */
+    public static final String UK_POST_CODE_REGEX = "([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})"
+            + "|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\\s?[0-9][A-Za-z]{2})";
+    @Id
+    @Getter(AccessLevel.NONE)
+    @Access(AccessType.PROPERTY)
```
Comment (YetAnotherMatt) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java]: Why have we got `@Access(AccessType.PROPERTY)`? I think we should be using the getter instead

File: api/src/main/resources/application.yml [hunk 11]
```diff
@@ -23,3 +23,5 @@ spring:
       default-property-inclusion: NON_NULL
     flyway:
         baseline-on-migrate: true
+server:
+    port: 8090
```
Comment (YetAnotherMatt) [inline api/src/main/resources/application.yml]: do we need this? We only need it to be 8090 when running locally, don't we?

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/AddressTestDataFactory.java (hunk 6 above)
Comment (YetAnotherMatt) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/AddressTestDataFactory.java]: Yes, lets remove the PMD rule in test

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+public class Address {
+
+    /**
+     * Regex for matching UK postcodes matching BS7666 format.
+     * { @see https://www.gov.uk/government/publications/bulk-data-transfer-for-sponsors-xml-schema }
+     * { @see https://stackoverflow.com/questions/164979/uk-postcode-regex-comprehensive }
+     */
+    public static final String UK_POST_CODE_REGEX = "([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})"
+            + "|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\\s?[0-9][A-Za-z]{2})";
+    @Id
```
Comment (YetAnotherMatt) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java]: if we're editing this anyway, blank line between the constant and @Id :)

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java (hunk 9 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/entity/Address.java]: Added reference.

File: api/src/main/resources/application.yml (hunk 11 above)
Comment (dwybourn) [reply api/src/main/resources/application.yml]: Removed accidental checkin
